spec/spec_helpers/worker_helper.rb

```diff
@@ -4,8 +4,8 @@ def process_one_job
       # The async worker creates a reactor and all the tests are run within a
       # reactor (see the config.around(:each) call on spec_helper.rb). To avoid
       # the worker creating a nested reactor inside the one already running, we
-      # run Worker.work in its own thread.
-      Thread.new { ThreeScale::Backend::Worker.work(one_off: true) }.join
+      # run Worker.work under a new reactor
```

Comment:
You're right, it doesn't create a new reactor as I thought. It just waits until the given task finishes.

> So it is not a new reactor. In fact it is within the existing reactor. And that's why > it would fail in a new thread. Do I get it right

Before the change, it failed because it was creating a new thread that scheduled new tasks, under the same [global reactor](https://github.com/3scale/apisonator/blob/72977e834c311eb592d478aa74d2f44428724a30/spec/spec_helper.rb#L63) being used by the main thread that runs the test. So different threads were adding tasks to the same reactor, which switches between those tasks, therefor sharing tasks across threads.

It works now because there's no new thread. The `Sync` call ensures it waits, to keep the old behavior, when we also waited for the thread to finish (`.join`).

CC @mayorova

Reply:
Fixed